src/System.Collections.Immutable/tests/ImmutableArrayTest.cs

```diff
@@ -1129,298 +1642,509 @@ public void ConcatEdgeCases()
         }
 
         [Fact]
-        public void IStructuralEquatable_Equals_NullComparerNullUnderlyingArray_Works()
+        public void IStructuralEquatableEqualsNullComparerInvalid()
         {
-            IStructuralEquatable equatable = new ImmutableArray<int>();
-            Assert.True(equatable.Equals(equatable, null));
-            Assert.False(equatable.Equals(ImmutableArray.Create(1, 2, 3), null));
-            Assert.True(equatable.Equals(new ImmutableArray<int>(), null));
-            Assert.Throws<NullReferenceException>(() => equatable.Equals(null, null));
+            // This was not fixed for compatability reasons. See https://github.com/dotnet/corefx/issues/13410
+            Assert.Throws<NullReferenceException>(() => ((IStructuralEquatable)ImmutableArray.Create(1, 2, 3)).Equals(ImmutableArray.Create(1, 2, 3), comparer: null));
+            Assert.Throws<NullReferenceException>(() => ((IStructuralEquatable)s_emptyDefault).Equals(other: null, comparer: null));
         }
 
-        [Fact]
-        public void StructuralEquatableGetHashCodeDefault()
+        [Theory]
+        [MemberData(nameof(IStructuralEquatableGetHashCodeData))]
+        public void IStructuralEquatableGetHashCode(IEnumerable<int> source, IEqualityComparer comparer)
         {
-            IStructuralEquatable defaultImmArray = s_emptyDefault;
-            Assert.Equal(0, defaultImmArray.GetHashCode(EqualityComparer<int>.Default));
+            var array = source.ToImmutableArray();
+            int expected = ((IStructuralEquatable)source.ToArray()).GetHashCode(comparer);
+            Assert.Equal(expected, ((IStructuralEquatable)array).GetHashCode(comparer));
         }
 
-        [Fact]
-        public void StructuralEquatableGetHashCode()
+        public static IEnumerable<object[]> IStructuralEquatableGetHashCodeData()
         {
-            IStructuralEquatable emptyArray = new int[0];
-            IStructuralEquatable emptyImmArray = s_empty;
-            IStructuralEquatable array = new int[3] { 1, 2, 3 };
-            IStructuralEquatable immArray = ImmutableArray.Create(1, 2, 3);
+            var enumerables = Int32EnumerableData()
+                .Select(array => array[0])
+                .Cast<IEnumerable<int>>();
 
-            Assert.Equal(emptyArray.GetHashCode(EqualityComparer<int>.Default), emptyImmArray.GetHashCode(EqualityComparer<int>.Default));
-            Assert.Equal(array.GetHashCode(EqualityComparer<int>.Default), immArray.GetHashCode(EqualityComparer<int>.Default));
-            Assert.Equal(array.GetHashCode(EverythingEqual<int>.Default), immArray.GetHashCode(EverythingEqual<int>.Default));
+            return SharedComparers<int>()
+                .OfType<IEqualityComparer>()
+                .Except(new IEqualityComparer[] { null })
```

Comment:
Self-nit: I think all of the `Except(new[] { null })`s may not be necessary. I found out (while writing this PR) what `OfType` does not include nulls.